alien.1
.TH alien 1L "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
alien \- Convert or install an alien binary package
.SH SYNOPSIS
\fBalien\fP [\fB--to-deb\fP] [\fB--patch=file\fP] [\fBoptions\fP] \fBfile\fP [\fB...\fP]
.br
\fBalien\fP --to-rpm [\fBoptions\fP] \fBfile\fP [\fB...\fP]
.br
\fBalien\fP --to-tgz [\fBoptions\fP] \fBfile\fP [\fB...\fP]
.br
\fBalien\fP --to-slp [\fBoptions\fP] \fBfile\fP [\fB...\fP]
.SH DESCRIPTION
To build Debian packages, 
.I alien
is simply called with a parameter giving the 
name of the alien package to be converted. A look up will then be done in
.B /var/lib/alien/packagename*.diff.gz
and then in
.B /usr/lib/alien/patches/packagename*.diff.gz .
If such a diff exists then the alien package is unpacked and the patch is
applied to debianize the package. If there is no such diff then 
.I alien 
will attempt to automatically debianize the package. After this, alien will 
build the debian binary package, and it will be saved to the current 
directory.
.PP
To build Red Hat packages, alien must be called with the --to-rpm parameter,
and the name of the alien package to be converted. 
.I Alien 
will then generate a spec file, and call rpm to build the package.
.PP
This tool probably needs to be run as superuser. Make sure that there is
enough room in the current directory since
.I alien
will build the package in that location.
.SH ALIEN PACKAGE FORMATS
.I Alien 
can input and output packages in the Red Hat .rpm format, the Debian .deb 
format, the Stampede .slp format, the Slackware .tgz format and the
generic .tar.gz format. 
.PP
For converting from and to .rpm format the Red Hat Package Manager must be 
installed (See
.B rpm (8)
).
.PP
For converting to (but not from) .deb format, the gcc, make, debmake, dpkg-dev,
and dpkg packages must be installed. 
.PP
Note that for the .tar.gz format, 
.I alien 
will simply generate a .deb or .rpm package that has the same files in it as 
are in the tar file. This only works well if the tar file has precompiled 
binaries in it in a standard linux directory tree. Do NOT run alien on tar 
files with source code in them, unless you want this source code to be 
installed in your root directory when you install the package!
.SH OPTIONS
.TP
.I file [...]
The list of files to convert.
.TP
.I -d, --to-deb
The output package will be a debian package. This is the default.
.TP
.I -r, --to-rpm
The output package will be a rpm package.
.TP
.I -t, --to-tgz
The output will be a slackware tgz package.
.TP
.I --to-slp
The output will be a stampede slp package.
.TP
.I -i, --install
Automatically install the output package, and remove the package file
after it has been installed.
.TP
.I -g, --generate
Generate a temporary directory suitable for building a package from, but do
not actually create the package. This is useful if you want to move files
around in the package before building it. The package can be built from this
temporary directory by running "debian/rules binary", if you were creating a
Debian package, or by running "rpm -bb <packagename>.spec" if you were
creating a Red Hat package.
.TP
.I -s, --single
Like -g, but do not generate the packagename.orig directory. This is only
useful when you are very low on disk space and are generating a debian
package.
.TP
.I --patch=patch
Specify the patch to be used instead of automatically looking the patch up
in
.B /var/lib/alien/ .
This is only allowed with --to-deb.
.TP
.I --nopatch
Do not use any patch files.
.TP
.I --description=desc
Specifiy a description for the package. This can only be used when
converting from slackware tgz packages, which lack descriptions. If it is
not used when converting from slackware tgz packages, a generic description
is used.
.TP
.I -c, --scripts
Try to convert the scripts that are meant to be run when the
package is installed and removed. Use this with caution, becuase these
scripts might be designed to work on a system unlike your own, and could
cause problems. It is reccommended that you examine the scripts by hand
and check to see what they do before using this option.
.TP
.I -k, --keep-version
By default, alien adds one to the minor version number. If this option
is given, alien will not do this.
.TP
.I -h, --help
Display a short usage summary.
.SH EXAMPLES
Here are some examples of the use of alien:
.TP
.I alien --to-deb package.rpm
Convert the package.rpm into a package.deb
.TP
.I alien --to-rpm package.deb
Convert the package.deb into a package.rpm
.TP
.I alien -i package.rpm
Convert the package.rpm into a package.deb (converting to a .deb package is
default, so you need not specify --to-deb), and install the generated
package.
.SH ENVIRONMENT
Alien recognizes the following environemnt variables:
.TP
.I RPMBUILDOPT
Options to pass to rpm when it is building a package.
.TP
.I RPMINSTALLOPT
Options to pass to rpm when it is installing a package.
.SH LIMITATIONS   
When running alien on a tar file, all files in /etc in are assumed to be
configuration files.
.PP
Alien does not account for differences in configuration between different
linux distributions. So don't use it to replace something essential like
sysvinit. You could destroy your system by doing so. In general, if you
can't uninstall the package without breaking your system, don't try to
replace it with an alien version.
.PP
If alien is not run as root, the files in the generated package will have
incorrect owners and permissions.
.SH AUTHOR
Alien was written by Christoph Lameter.
.br
<clameter@debian.org>
.PP
Deb to rpm conversion code was taken from the Martian program by
Randolph Chung.
.br
<rc42@cornell.edu>
.PP
Alien has been extensively rewritten and is now maintained by Joey Hess.
.br
<joeyh@debian.org>


CHANGES
alien (5.0) unstable; urgency=low

  * Added some cautions to man page about not using alien to replace
    important packages.
  * This version will be released to the linux community at large, not just
    debian.
  * Added a README file.

 -- Joey Hess <joeyh@master.debian.org>  Wed, 7 May 1997 16:02:53 -0400

alien (4.3) unstable; urgency=low

  * Fixed a bug that made alien choke on packages that had 0 for their
    release or version number.
  * Removed obsolete alien.sh from the source package.

 -- Joey Hess <joeyh@master.debian.org>  Wed, 7 May 1997 15:17:33 -0400

alien (4.2) unstable; urgency=low

  * When installing deb file, alien will use --no-force-overwrite
    The idea behind this is to make it difficult to trash your debian system
    by installing alien packages that overwrite files in it. This only works 
    if you use alien --install, not if you install the resulting .deb file by
    hand. This is a temporary fix until dpkg has --force-overwrite turned
    off by default.

 -- Joey Hess <joeyh@master.debian.org>  Wed, 16 Apr 1997 17:03:27 -0400

alien (4.1) unstable; urgency=low

  * If a package has underscores in it's name and is being converted to deb
    format, change the underscores to dashes. Thanks to  
    Robert Coie <rac@mata.intrigue.com>
  * Strip out any other disallowed characters in package name when
    converting to deb.

 -- Joey Hess <joeyh@master.debian.org>  Fri, 4 Apr 1997 20:12:07 -0500

alien (4.00) unstable; urgency=low

  * Added support for converting deb to rpm, based on the "martian" program
    by Randolph Chung <rc42@cornell.edu>.
  * Huge reorganization, rewrite, and code cleanup.
  * As a side effect of the above, alien can also convert tgz into rpm now.
  * Use rpm -qcp to list conffiles from rpm files.
  * Reworked how conffiles are found from tar files.
  * Removed --noinstall option, changed alien to default to not installing
    generated packages, added --install option to make it install packages.
  * Removed --nopatch/--auto option, and changed patch lookup behavior. See
    the man page for an explination of the new behavior.
  * Now Recommends: dpkg-dev and make and Suggests: patch
  * Dropped feature of making slackware package install scripts into
    postinst. If there is any demand, I'll try to work this back into the
    program.
  * Rewrote most of the man page.
  * Fixed the version number so there are two digets in the minor revision 
    number.

 -- Joey Hess <joeyh@master.debian.org>  Sat, 29 Mar 1997 21:49:30 -0500

alien (3.2) frozen unstable; urgency=low

  * Fixed bug that was preventing alien from figuring out the version number
    of a tar file. (#8284)
  * Man page fixes, related to #8284.
  * Fixed a bug, present since 3.0, which was messing up conffile detection.
  * Removed README.debian. (It just pointed users to the man page, which is
    rather pointless.)

 -- Joey Hess <joeyh@master.debian.org>  Mon, 24 Mar 1997 14:03:04 -0500

alien (3.1) unstable; urgency=low

  * Force package names to lowercase.
  * Fixed --noinstall option.

 -- Joey Hess <joeyh@master.debian.org>  Mon, 17 Mar 1997 18:35:46 -0500

alien (3.00) unstable; urgency=low

  * Rewrote alien in perl.
  * Command line options can be specified in any order.
  * -ppatch will no longer work. use --patch=<patchfile> instead.
  * Improved usage help.
  * Include extended description from rpm, if available.
  * Fixed a bug with descriptions/summaries/changelogs/copyrights on rpms
    that contained the '/' character crashing alien.
  * Increment revision number of rpms when they are alianized.
  * Recommends: cpio becuase it is needed for rpm2cpio extraction.
  * Patch files do not have to be compressed.
  * Routine update of debian/rules:
    Clean up junk files in subdirs.

 -- Joey Hess <joeyh@master.debian.org>  Fri, 7 Mar 1997 16:08:02 -0500

alien (2.82) unstable; urgency=low

  * Use CHANGELOGTEXT, not CHANGELOG, when querying rpm files for
    changelogs. Fixes bug #7445.
  * Modification to work with rpm verison 2.3.7. This breaks compatability
    with previous versions of rpm. Modified control file to reflect this.

 -- Joey Hess <joeyh@master.debian.org>  Tue, 25 Feb 1997 21:09:18 -0500

alien (2.81) unstable; urgency=low

  * Corrected maintainer in debian/control.
  * Routine update of debian/rules:
    Modifications for multiple binary package support.

 -- Joey Hess <joeyh@master.debian.org>  Fri, 7 Feb 1997 22:25:40 -0500

alien (2.80) unstable; urgency=low

  * Added a patch file for Applixware.
  * I'm looking for patch files for other commercial software, to add to the
    package. If you own commerical linux software and would like to
    contribute a patch, please contact me.

  * Generated debian/rules files will no longer pass package name to debstd
    (And neither does the debian/rules file for this package.)
  * Added -s switch for low disk space situations.
  * Added -i switch to build a package but not install it.
  * Added long options (--option), as I can never remember the short ones.
  * Fixed bug in guessing name of patch file to use.
  * Fixed bug that would not let you specify a patch file in the current
    directory, or a relative path to a patch file.
  * Don't use /etc/rpmrc as an indiciation of whether rpm is present, as 
    this is a conffile, and might be deleted. Test for actual rpm binary.
  * More friendly error message if patch file is not found, suggesting that
    you try -n option.
  * Rewrote the code that figures out information about the rpm files so it
    uses rpm --queryformat to determine everything.
  * If a rpm has a changelog, add it to the end of debian/changelog.
  * Add summary to the Description: field of generated control files.
  * Add info from a rpm's copyright field to to the copyright file.

 -- Joey Hess <joeyh@master.debian.org>  Thu, 30 Jan 1997 20:35:10 -0500

alien (2.79) unstable; urgency=low

  * New maintainer.
  * Updated man page with new author info.

alien.pl
#!/usr/bin/perl -w

=head1 NAME

alien - Convert or install an alien binary package

=head1 SYNOPSIS

 alien [--to-deb] [--to-rpm] [--to-tgz] [--to-slp] [options] file [...]

=head1 DESCRIPTION

B<alien> is a program that converts between Redhat rpm, Debian deb,
Stampede slp, Slackware tgz, and Solaris pkg file formats. If you want to
use a package from another linux distribution than the one you have
installed on your system, you can use alien to convert it to your preferred
package format and install it. It also supports LSB packages.

=head1 WARNING

Despite the high version number, alien is still (and will probably always
be) rather experimental software. It's been under development for many
years now, but there are still many bugs and limitations.

Alien should not be used to replace important system packages, like
init, libc, or other things that are essential for the functioning of
your system. Many of these packages are set up differently by the
different distributions, and packages from the different distributions
cannot be used interchangeably. In general, if you can't remove a
package without breaking your system, don't try to replace it with an
alien version.

=head1 PACKAGE FORMAT NOTES

=over 4

=item rpm

For converting to and from rpm format the Red Hat Package Manager must be
installed.

=item lsb

To convert from lsb packages, the Red Hat Package Manager must be installed.
Unlike the other package formats, alien can handle the depenendencies of
lsb packages if the destination package format supports dependencies. Note
that this means that the package generated from a lsb package will depend on
a package named "lsb" -- your distribution should provide a package by that
name, if it is lsb compliant. The scripts in the lsb package will be converted
by default as well.

To generate lsb packages, the Red Hat Package Manager must be installed,
and alien will use by preference a program named lsb-rpm, if it exists.
No guarantees are made that the generated lsb packages will be fully LSB
compliant, and it's rather unlikely they will unless you build them in the
lsbdev environment.

=item deb

For converting to (but not from) deb format, the gcc, make, debmake,
dpkg-dev, and dpkg packages must be installed.

=item tgz

Note that when converting from the tgz format, B<alien> will simply generate an
output package that has the same files in it as are in the tgz file. This
only works well if the tgz file has precompiled binaries in it in a
standard linux directory tree. Do NOT run alien on tar files with source
code in them, unless you want this source code to be installed in your root
directory when you install the package!

=item pkg

To manipulate packages in the Solaris pkg format (which is really the SV 
datastream package format), you will need the Solaris pkginfo and pkgtrans
tools.

=back

=head1 OPTIONS

Alien will convert all the files you pass into it into all the output types
you specify. If no output type is specified, it defaults to converting to
deb format.

=over 4

=item file [...]

The list of package files to convert.

=item B<-d>, B<--to-deb>

Make debian packages. This is the default.

=item B<-r>, B<--to-rpm>

Make rpm packages.

=item B<-t>, B<--to-tgz>

Make tgz packages.

=item B<--to-slp>

Make slp packages.

=item B<-p>, B<--to-pkg>

Make Solaris pkg packages.

=item B<-i>, B<--install>

Automatically install each generated package, and remove the package file
after it has been installed.

=item B<-g>, B<--generate>

Generate a temporary directory suitable for building a package from, but do
not actually create the package. This is useful if you want to move files
around in the package before building it. The package can be built from
this temporary directory by running "debian/rules binary", if you were creating
a Debian package, or by running "rpm -bb <packagename>.spec" if you were
creating a Red Hat package.

=item B<-s>, B<--single>

Like B<-g>, but do not generate the packagename.orig directory. This is only
useful when you are very low on disk space and are generating a debian
package.

=item B<--patch=>I<patch>

Specify the patch to be used instead of automatically looking the patch up
in B</var/lib/alien>. This has no effect unless a debian package is being
built.

=item B<--nopatch>

Do not use any patch files.

=item B<--description=>I<desc>

Specifiy a description for the package. This only has an effect when
converting from the tgz package format, which lacks descriptions.

=item B<-c>, B<--scripts>

Try to convert the scripts that are meant to be run when the
package is installed and removed. Use this with caution, becuase these
scripts might be designed to work on a system unlike your own, and could
cause problems. It is recommended that you examine the scripts by hand
and check to see what they do before using this option.

This is enabled by default when converting from lsb packages.

=item B<-T>, B<--test>

Test the generated packages. Currently this is only supported for debian
packages, which, if lintian is installed, will be tested with lintian and
lintian's output displayed.

=item B<-k>, B<--keep-version>

By default, alien adds one to the minor version number of each package it
converts. If this option is given, alien will not do this.

=item B<--fixperms>

Sanitize all file owners and permissions when building a deb. This may be
useful if the original package is a mess. On the other hand, it may break
some things to mess with their permissions and owners to the degree this does,
so it defaults to off. This can only be used when converting to debian
packages.

=item B<-h>, B<--help>

Display a short usage summary.

=back

=head1 EXAMPLES

Here are some examples of the use of alien:

=over 4

=item alien --to-deb package.rpm

Convert the package.rpm into a package.deb

=item alien --to-rpm package.deb

Convert the package.deb into a package.rpm

=item alien -i package.rpm

Convert the package.rpm into a package.deb (converting to a .deb package is
default, so you need not specify --to-deb), and install the generated
package.

=item alien --to-deb --to-rpm --to-tgz --to-slp foo.deb bar.rpm baz.tgz

Creates 9 new packages. When it is done, foo bar and baz are available in
all 4 package formats.

=back

=head1 ENVIRONMENT

Alien recognizes the following environemnt variables:

=over 4

=item RPMBUILDOPT

Options to pass to rpm when it is building a package.

=item RPMINSTALLOPT

Options to pass to rpm when it is installing a package.

=item EMAIL

If set, alien assumes this is your email address. Email addresses are included
in generated debian packages.

=back

=head1 NOTES

When using alien to convert a tgz package, all files in /etc in are assumed
to be configuration files.

If alien is not run as root, the files in the generated package will have
incorrect owners and permissions.

=head1 AUTHOR

Alien was written by Christoph Lameter, B<<clameter@debian.org>>.

deb to rpm conversion code was taken from the Martian program by
Randolph Chung, B<<tausq@debian.org>>.

The Solaris pkg code was written by Mark A. Hershberger B<<mah@everybody.org>>.

Alien has been extensively rewritten (3 times) and is now maintained by
Joey Hess, B<<joeyh@debian.org>>.

=head1 COPYRIGHT

Alien may be copied amd modified under the terms of the GNU General Public
License.

=cut

use strict;
use lib '.'; # For debugging, removed by Makefile.
use Getopt::Long;
use Alien::Package::Deb;
use Alien::Package::Rpm;
use Alien::Package::Tgz;
use Alien::Package::Slp;
use Alien::Package::Pkg;
use Alien::Package::Lsb;

# Returns a list of directories to search for patches.
sub patchdirs {
	return '/var/lib/alien',"/usr/share/alien/patches";
}

# Display alien's version number.
sub version {
	my $version_string='unknown'; # VERSION_AUTOREPLACE done by Makefile, DNE
	print "Alien version $version_string\n";
	exit;
}

# Display usage help.
sub usage {
	print STDERR <<EOF;
Usage: alien [options] file [...]
  file [...]                Package file or files to convert.
  -d, --to-deb              Generate a Debian deb package (default).
     Enables the following options:
       --patch=<patch>      Specify patch file to use instead of automatically
                            looking for patch in /var/lib/alien.
       --nopatch	    Do not use patches.
       --single             Like --generate, but do not create .orig
                            directory.
       --fixperms           Munge/fix permissions and owners.
       --test               Test generated packages with lintian.
  -r, --to-rpm              Generate a RedHat rpm package.
      --to-slp              Generate a Stampede slp package.
  -l, --to-lsb              Generate a LSB package.
  -t, --to-tgz              Generate a Slackware tgz package.
     Enables the following option:
       --description=<desc> Specify package description.
  -p, --to-pkg              Generate a Solaris pkg package.
  -i, --install             Install generated package.
  -g, --generate            Unpack, but do not generate a new package.
  -c, --scripts             Include scripts in package.
  -k, --keep-version        Do not change version of generated package.
  -h, --help                Display this help message.
  -v, --version		    Display alien's version number.

EOF
	exit 1;
}

# Start by processing the parameters.
my (%destformats, $generate, $install, $single, $scripts, $patchfile,
    $nopatch, $tgzdescription, $keepversion, $fixperms, $test);

# Bundling is nice anyway, and it is required or Getopt::Long will confuse
# -T and -t.
Getopt::Long::Configure("bundling");

GetOptions(
	"to-deb|d", sub { $destformats{deb}=1 },
	"to-rpm|r", sub { $destformats{rpm}=1 },
	"to-lsb|l", sub { $destformats{lsb}=1 },
	"to-tgz|t", sub { $destformats{tgz}=1 },
	"to-slp",   sub { $destformats{slp}=1 },
	"to-pkg|p", sub { $destformats{pkg}=1 },
	"test|T", \$test,
	"generate|g", \$generate,
	"install|i", \$install,
	"single|s", sub { $single=1; $generate=1 },
	"scripts|c", \$scripts,
	"patch|p=s", \$patchfile,
	"nopatch", \$nopatch,
	"description=s", \$tgzdescription,
	"keep-version|k", \$keepversion,
	"fixperms", \$fixperms,
	"help|h", \&usage,
	"version|v", \&version,
) || usage();

# Default to deb conversion.
if (! %destformats) {
	$destformats{deb}=1;
}

# A few sanity checks.
if (($generate || $single) && $install) {
	die "You can not use --generate or --single with --install.\n";
}
if (($generate || $single) && keys %destformats > 1) {
	die "--generate and --single may only be used when converting to a single format.\n";
}
if ($patchfile && ! -f $patchfile) {
	die "Specified patch file, \"$patchfile\" cannot be found.\n";
}
if ($patchfile && $nopatch) {
	die "The options --nopatch and --patchfile cannot be used together.\n";
}
unless (@ARGV) {
	print STDERR "You must specify a file to convert.\n\n";
	usage();
}

# Check alien's working anvironment.
if (! -w '.') {
	die("Cannot write to current directory. Try moving to /tmp and re-running alien.\n");
}
if ($> ne 0) {
	if ($destformats{deb} && ! $generate && ! $single) {
		die "Must run as root to convert to deb format (or you may use fakeroot).\n";
	}
	print STDERR "Warning: alien is not running as root!\n";
	print STDERR "Ownerships of files in the generated packages will probably be messed up.\n";
}

foreach my $file (@ARGV) {
	if (! -f $file) {
		die "File \"$file\" not found.\n";
	}

	# Figure out what kind of file this is.
	my $package;

	# Check lsb before rpm, since lsb packages are really just
	# glorified rpms.
	if (Alien::Package::Lsb->checkfile($file)) {
		$package=Alien::Package::Lsb->new(filename => $file);
	}
	elsif (Alien::Package::Rpm->checkfile($file)) {
		$package=Alien::Package::Rpm->new(filename => $file);
	}
	elsif (Alien::Package::Deb->checkfile($file)) {
		$package=Alien::Package::Deb->new(filename => $file);
	}
	elsif (Alien::Package::Tgz->checkfile($file)) {
		$package=Alien::Package::Tgz->new(filename => $file);
		$package->description($tgzdescription) if defined $tgzdescription;
	}
	elsif (Alien::Package::Slp->checkfile($file)) {
		$package=Alien::Package::Slp->new(filename => $file);
	}
	elsif (Alien::Package::Pkg->checkfile($file)) {
		$package=Alien::Package::Pkg->new(filename => $file);
	}
	else {
		die "Unknown type of package, $file.\n";
	}

	$package->usescripts($scripts) unless $package->usescripts;

	# Increment release.
	unless (defined $keepversion) {
		$^W=0; # Shut of possible "is not numeric" warning.
		$package->release($package->release + 1);
		$^W=1; # Re-enable warnings.
	}

	foreach my $format (keys %destformats) {
		# Skip conversion if package is already the correct format.
		if ($package->origformat ne $format) {
			# Only unpack once.
			if ($package->unpacked_tree) {
				$package->cleantree;
			}
			else {
				$package->unpack;
			}
			
			# Mutate package into desired format.
			bless($package, "Alien::Package::".ucfirst($format));
		
			# Make .orig.tar.gz directory?
			if ($format eq 'deb' && ! $single && $generate) {
				# Make .orig.tar.gz directory.
				system("cp", "-fa", "--", $package->unpacked_tree, $package->unpacked_tree.".orig") == 0
					or die "cp -fa failed";
			}
	
			# See if a patch file should be used.
			if ($format eq 'deb' && ! $nopatch) {
				if (defined $patchfile) {
					$package->patchfile($patchfile)
				}
				else {
					$package->patchfile($package->getpatch(patchdirs()));
				}
			}

			$package->fixperms($fixperms);
			
			$package->prep;
			
			# If generating build tree only, stop here with message.
			if ($generate) {
				if ($format eq 'deb' && ! $single) {
					print "Directories ".$package->unpacked_tree." and ".$package->unpacked_tree.".orig prepared.\n"
				}
				else {
					print "Directory ".$package->unpacked_tree." prepared.\n";
				}
				# Make sure $package does not wipe out the
				# directory when it is destroyed.
				$package->unpacked_tree('');
				exit;
			}
			
			my $newfile=$package->build;
			if ($test) {
				my @results = $package->test($newfile);
				if (@results) {
					print "Test results:\n";
					print "\t$_\n" foreach @results;
				}
			}
			if ($install) {
				$package->install($newfile);
				unlink $newfile;
			}
			else {
				# Tell them where the package ended up.
				print "$newfile generated\n";
			}
		}
		elsif ($install) {
			# Don't convert the package, but do install it.
			$package->install($file);
			# Note I don't unlink it. I figure that might annoy
			# people, since it was an input file.
		}

		$package->revert;
	}
}
